### PR TITLE
feat-pagerduty - added conversion for pagerduty plugin

### PR DESCRIPTION
**Pull request by vanisrikanithi**
pagerduty go-conversion:
Added "Pagerduty " support to go-convert specific to jenkins migration.

Pager Duty Jenkins Trace: 
[test-pager-duty.json](https://github.com/user-attachments/files/17782636/test-pager-duty.json)

go run main.go jenkinsjson --downgrade ~/go/test-pager-duty.json

pagerduty step:
```
- step:
    identifier: pagerduty51d22b
    name: Pagerduty
    spec:
      image: plugins/pagerduty
      settings:
        dedup_key: E54EC853A59A3815EF3632D5F854CF26
        incident_severity: critical
        incident_source: test-pager-duty
        incident_summary: Build Failed for test-pager-duty
        job_status: failure
        log_level: debug
        resolve: 'false'
        routing_key: a666ad1326f34605d06c0dbd4d87c1cb
    timeout: ''
    type: Plugin
```

pagerdutyChangeEvent step:
```
- step:
    identifier: pagerdutychangeevent499434
    name: Pagerduty_Change_Event
    spec:
      image: plugins/pagerduty
      settings:
        custom_details: '{"buildNumber":"22","jobName":"test-pager-duty","jobURL":"http://localhost:8080/job/test-pager-duty/22/"}'
        incident_source: ''
        incident_summary: Job test-pager-duty completed with status SUCCESS
        log_level: debug
        routing_key: a666ad1326f34605d06c0dbd4d87c1cb
        create_change_event: "true"
    timeout: ''
    type: Plugin
```

### CODE REVIEW

**Comment by vanisrikanithi** [comment]
@Ompragash Added the pagerduty and pagerdutyChangeEvent support to go-convert, Could you please review it and confirm the changes?

Thanks,

**Comment by vanisrikanithi** [comment]
@Ompragash Addressed all the above comments, could you please review them and confirm?

File: convert/jenkinsjson/json/pager_duty.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+	convertPagerduty := &harness.Step{
+		Id:   SanitizeForId(node.SpanName, node.SpanId),
+		Name: "Pagerduty",
+		Type: "plugin",
+		Spec: &harness.StepPlugin{
+			Image: "plugins/pagerduty",
+			With: map[string]interface{}{
+				"log_level":         "info",
+				"routing_key":       routingKey,
```
Comment (Ompragash) [inline convert/jenkinsjson/json/pager_duty.go:27]:
```suggestion
				"routing_key":       <+input>,
```

changed it to get user input as routing key is a secret

**Review by Ompragash — COMMENTED** [review]
test is failing with `convert/jenkinsjson/json/pager_duty.go:27:26: syntax error: unexpected <, expected expression`

**Comment by vanisrikanithi** [comment]
@Ompragash, Please find the [LINK](https://app.harness.io/ng/account/gCoPSwHxS7ipOgx2iA9tOQ/module/ci/orgs/default/projects/jenkinstoharness/pipelines/vani_pagerduty_plugin/executions/Dh4PxviSTqaFhs3z5ffE7g/pipeline) for pagerduty conversion test pipeline.

**Comment by vanisrikanithi** [comment]
@Ompragash, Added IntegrationKey for PagerdutyChangeEvent, please find the below test pipelines.

[Pagerduty Test Release Pipeline](https://app.harness.io/ng/account/gCoPSwHxS7ipOgx2iA9tOQ/module/ci/orgs/default/projects/jenkinstoharness/pipelines/VaniPagerdutydronereleasepipeline/executions/wqXaLuutSpqAybbjWZkqqg/pipeline)
[Pagerduty test pipeline](https://app.harness.io/ng/account/gCoPSwHxS7ipOgx2iA9tOQ/module/ci/orgs/default/projects/jenkinstoharness/pipelines/vani_pagerduty_plugin/executions/FZ9rMTI8Rqy2Yhc3x2M66Q/pipeline)

Could you please review them and confirm the changes?

File: convert/jenkinsjson/json/pager_duty.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+	convertPagerduty := &harness.Step{
+		Id:   SanitizeForId(node.SpanName, node.SpanId),
+		Name: "Pagerduty",
+		Type: "plugin",
+		Spec: &harness.StepPlugin{
+			Image: "plugins/pagerduty",
+			With: map[string]interface{}{
+				"log_level":         "info",
+				"routing_key":       <+input>,
```
Comment (Ompragash) [inline convert/jenkinsjson/json/pager_duty.go:27]:
```suggestion
				"routing_key":       "<+input>",
```

Comment (Ompragash) [reply convert/jenkinsjson/json/pager_duty.go:27]: routing keys are secrets, so let's keep it as `<+input>`

File: convert/jenkinsjson/json/pager_duty.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+	convertPagerdutyChangeEvent := &harness.Step{
+		Name: "Pagerduty_Change_Event",
+		Type: "plugin",
+		Id:   SanitizeForId(node.SpanName, node.SpanId),
+		Spec: &harness.StepPlugin{
+			Image: "plugins/pagerduty",
+			With: map[string]interface{}{
+				"log_level":           "info",
+				"routing_key":         routingkey,
```
Comment (Ompragash) [inline convert/jenkinsjson/json/pager_duty.go:66]: same here

File: convert/jenkinsjson/json/pager_duty.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
```
Comment (Ompragash) [inline convert/jenkinsjson/json/pager_duty.go:66]:
```suggestion
				"routing_key":         "<+input>",
```

File: convert/jenkinsjson/json/pager_duty_test.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+				t.Errorf("ConvertPagerDuty() mismatch (-want +got):\n%s", diff)
+			}
+		})
+	}
+}
+
+func TestConvertPagerDutyChangeEvent(t *testing.T) {
+
+	var tests []runner
+	tests = append(tests, prepare(t, "/pagerdutyChangeEvent/pagerdutyChangeEvent_snippet", &harness.Step{
```
Comment (Ompragash) [inline convert/jenkinsjson/json/pager_duty_test.go:46]: test case is failing here with the below - just update the snippet json file name
```
ok  	github.com/drone/go-convert/convert/jenkinsjson	0.013s	coverage: 25.4% of statements
--- FAIL: TestConvertPagerDutyChangeEvent (0.00s)
    base_test.go:27: failed to read JSON file: open /drone/src/convert/jenkinsjson/convertTestFiles/pagerdutyChangeEvent/pagerdutyChangeEvent_snippet.json: no such file or directory
Missing 'delegate' in parameterMap
No entries operations found in arguments
```

**Review by Ompragash — COMMENTED** [review]
ok  	github.com/drone/go-convert/convert/jenkinsjson	0.013s	coverage: 25.4% of statements
--- FAIL: TestConvertPagerDutyChangeEvent (0.00s)
    base_test.go:27: failed to read JSON file: open /drone/src/convert/jenkinsjson/convertTestFiles/pagerdutyChangeEvent/pagerdutyChangeEvent_snippet.json: no such file or directory
Missing 'delegate' in parameterMap
No entries operations found in arguments
2024/12/03 04:32:46 withCredentials unsupported symbol: not_real
FAIL
coverage: 46.8% of statements

**Pull request merged**